### PR TITLE
Add command for define-fun-rec and add to API

### LINKED ISSUE #839: define-fun-rec has no command (Bugzilla #620)
**Imported from Bugzilla**
Reporter: *Tim King*
Status: *NEW*
Severity: *normal*
Assigned to: *Andrew Reynolds*
Component: *{Internals} Parser infrastructure*
Milestone: *1.6*
Version: *master*
Platform: *PC*
OS: *Linux*

Original attachment names and IDs:
- _[datatypeList.smt2](http://cvc4.cs.stanford.edu/bugzilla_attachments/269_datatypeList.smt2)_ (define-fun-rec bug, ID 269)
- _[datatypeList-fix.smt2](http://cvc4.cs.stanford.edu/bugzilla_attachments/311_datatypeList-fix.smt2)_ (fixed syntax, uncommented rec-fun, ID 311)

On 2015-03-02 09:51:02 -0800, Tim King wrote:
> Comes from a bug reported by Pierre van de Laar.
> 
> 
> It looks like the define\-fun\-rec implementation is incomplete. There is no command type backing this command. (Or if there is one, I can't find it.) The Smt2.g parser thus cannot set the cmd flag when handling the command:
> Â  Â https://github.com/CVC4/CVC4/blob/master/src/parser/smt2/Smt2.g#L872-L904
> Thus cmd remains null when it returns into parser.cpp's nextCommand() and the local variable cmd is set to null:
> Â  Â https://github.com/CVC4/CVC4/blob/master/src/parser/smt2/Smt2.g#L872-L904
> This prematurely forces the parser to exit. Subsequent commands are not parsed.
> 
> If there are no objections, I am going to make define\-fun\-rec throw an exception at the beginning of parsing the command. This is unless a command line argument is provided so we can continue to test the partial implementation in the parser.
> 
> We can discuss what the define\-fun\-rec command should be at the meeting.

On 2015-03-02 09:52:46 -0800, Tim King wrote:
> Created attachment 269
> define\-fun\-rec bug
> 
> Pierre van de Laar's bug submitted to the cvc4\-bugs list.

On 2015-08-20 06:36:11 -0700, Andrew Reynolds wrote:
> Created attachment 311
> fixed syntax, uncommented rec\-fun

On 2015-08-20 06:36:38 -0700, Andrew Reynolds wrote:
> define\-fun\-rec is now supported in the SMT2 parser, but not in the API.  I've attached the updated version of attachment 269 in the correct syntax.

On 2016-12-14 14:39:05 -0800, Clark Barrett wrote:
> Is it still the case that define\-fun\-rec is not supported in the API?

On 2016-12-14 14:40:43 -0800, Andrew Reynolds wrote:
> Yes, unfortunately this is still the case.  It wouldn't be hard to do this, but this will take 3\-4 hours to do.

### PR DESCRIPTION
**Pull request by ajreynol**
This fixes #839, and is work towards #1316.

The pull request removes some of the (duplicated, hard-coded) code from Smt2.g into smt_engine.cpp.

It adds defineFunctionsRec to the API in smt engine.

It also slightly improves the input types of setUserAttribute in a few places, and fixes documentation on SmtEngine::defineFunction.

### CODE REVIEW

File: src/printer/smt2/smt2_printer.cpp [hunk 1]
```diff
@@ -1501,6 +1502,74 @@ static void toStream(std::ostream& out, const DefineFunctionCommand* c)
   out << ") " << type << " " << formula << ")";
 }
 
+static void toStream(std::ostream& out, const DefineFunctionRecCommand* c)
+{
+  const vector<Expr>& funcs = c->getFunctions();
+  const vector<vector<Expr> >& formals = c->getFormals();
+  out << "(define-fun";
+  if (funcs.size() > 0)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/printer/smt2/smt2_printer.cpp]: Yes, I'll fix

File: src/smt/command.cpp [hunk 2]
```diff
@@ -890,6 +890,94 @@ Command* DefineNamedFunctionCommand::clone() const {
...
+  {
+    d_commandStatus = new CommandFailure(e.what());
+  }
+}
+
+Command* DefineFunctionRecCommand::exportTo(
+    ExprManager* exprManager, ExprManagerMapCollection& variableMap)
+{
+  std::vector<Expr> funcs;
+  for (unsigned i = 0, size = d_funcs.size(); i < size; i++)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/command.cpp]: We don't have a unit test for ensures for-each work for Expr, so currently I don't try for-each for Expr.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

File: src/smt/smt_engine.h [hunk 3]
```diff
@@ -448,6 +453,32 @@ class CVC4_PUBLIC SmtEngine {
   /** is defined function */
   bool isDefinedFunction(Expr func);
 
+  /** Define functions recursive
+   *
+   * For each i, this contrains funcs[i] in the current context to be:
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/smt_engine.h]: Sorry missed this one

File: src/smt/command.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/smt/command.cpp]: Well, it would be a for-each over the vector, not the Expr but it is not very critical.

**Pull request merged**